### DISCUSSION TITLE
Documentation on custom export format implementation?

### DISCUSSION THREAD

**Discussion by naquad** [zim-desktop-wiki/zim-desktop-wiki#2572] [Q&A] [answered]
Hi,

I want to implement a custom markdown exporter. I haven't found any mentions of custom exporters (format + template) in the Wiki, tried to look through the code, but all of the functions are using hard-coded module names like `"zim.format." + canonical_name(format)` or stuff. It could be that I'm missing some magic, but some clarification would be greatly appreciated.

Thank you.

**Comment by jaap-karssenberg** [comment, marked as answer]
Yes, currently there is no plugin hook for formats. Also not planning to do
so. Long term I would like te reuse tools like pandoc.

For now just copy existing markdown module and patch in place to modify the
format.

Regards

Jaap

Op ma 12 feb. 2024 00:43 schreef Daniel ***@***.***>:

**Reply by naquad** [reply]
I see, ok then, that's what I'll do, thank you.

**Reply by naquad** [reply]
As a follow-up question, could you please give a word of advice to someone who's trying to change the structure of the output dir?

What I'm trying to do:
* Replace `_` with ` ` in the file names (changing `encode_filename` causes wreck and havoc and yields lots of unprocessed txt files that are actually pages)
* Move all attachments to a single directory (as a consequence change the duplicate file names, totally clueless, any tries to mess with the attachments_dir layout method causes overwritten files and broken links)

Thank you.

**Reply by introt** [reply]
see also https://gist.github.com/introt/fed672dfbea9df399610084be448326b - a plugin that monkey patches the existing exporter to inline images

**Comment by jaap-karssenberg** [comment]
For exporting you have an "exporter" object and a "layout" object. A
different layout for page names in the target folder can be done by
overloading the methods in a layout object. You should indeed not redefine
"encode_filename" as it is a helper function used in multiple places.
Better to define a modified copy of that function and call it from the
overloaded layout method.

Moving all attachments to a single folder is more tricky. And I can't tell
you how to do that exactly without more details on how to do that. For sure
you will have to modify the exporter object and not just the layout. Have a
look at the "export_attachments_iter()" method in the files exporter. You
can use "folder.new_file()" method to avoid conflicts (but make sure to
clear the folder first, otherwise running the same export multiple times
will not overwrite the old attachments, but keeps making copies).
Also in this case you need to think about the links to the attachments.
Once attachments are renamed, links can be broken. For this there is a
Linker object that should have knowledge about the naming scheme. However
sorting out the conflicts and understanding which copy is which original
file is up to you and non trivial I'm afraid.

Hope this helps,

Jaap

**Comment by naquad** [comment]
Thank you very much for the reply. After multiple attempts, I've made a
decision to make only minimal changes in the source code and use external
tools to change the output structure. Not as elegant as it could be but
easier and already done :)

вт, 20 февр. 2024 г. в 21:25, Jaap Karssenberg ***@***.***>:

> For exporting you have an "exporter" object and a "layout" object. A
> different layout for page names in the target folder can be done by
> overloading the methods in a layout object. You should indeed not redefine
> "encode_filename" as it is a helper function used in multiple places.
> Better to define a modified copy of that function and call it from the
> overloaded layout method.
>
> Moving all attachments to a single folder is more tricky. And I can't tell
> you how to do that exactly without more details on how to do that. For
> sure
> you will have to modify the exporter object and not just the layout. Have
> a
> look at the "export_attachments_iter()" method in the files exporter. You
> can use "folder.new_file()" method to avoid conflicts (but make sure to
> clear the folder first, otherwise running the same export multiple times
> will not overwrite the old attachments, but keeps making copies).
> Also in this case you need to think about the links to the attachments.
> Once attachments are renamed, links can be broken. For this there is a
> Linker object that should have knowledge about the naming scheme. However
> sorting out the conflicts and understanding which copy is which original
> file is up to you and non trivial I'm afraid.
>
> Hope this helps,
>
> Jaap
>
>
>
>
>
>
> On Mon, Feb 12, 2024 at 11:38 PM Daniel ***@***.***> wrote:
>
> > As a follow-up question, could you please give a word of advice to
> someone
> > who's trying to change the structure of the output dir?
> >
> > What I'm trying to do:
> >
> > - Replace _ with in the file names (changing encode_filename causes
> > wreck and havoc and yields lots of unprocessed txt files that are
> actually
> > pages)
> > - Move all attachments to a single directory (as a consequence change
> > the duplicate file names, totally clueless, any tries to mess with the
> > attachments_dir layout method causes overwritten files and broken links)
> >
> > Thank you.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/2572#discussioncomment-8447519>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAJQYHX74Q5ZAU2F6MMJZHDYTKKXVAVCNFSM6AAAAABDD5EW2CVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DINBXGUYTS>
>
> > .
> > You are receiving this because you commented.Message ID:
> >
> <zim-desktop-wiki/zim-desktop-wiki/repo-discussions/2572/comments/8447519@
> > github.com>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/2572#discussioncomment-8534343>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAE7FJMZTH3LTU4QRILENDYUT2BXAVCNFSM6AAAAABDD5EW2CVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DKMZUGM2DG>
> .
> You are receiving this because you authored the thread.Message ID:
> <zim-desktop-wiki/zim-desktop-wiki/repo-discussions/2572/comments/8534343@
> github.com>
>


-- 
Regards, Danil Fartushnii.
--------------------------------------------
Danil Fartushnii
Senior developer
Skype: dfartushnii
Jabber: ***@***.***
ICQ: 226555656

**Reply by brucewestfall** [reply]
Can you give some examples of what youve done and how you accomplished it?  If nothing els eit will be very helpful for the 'historical documents'

**Reply by naquad** [reply]
What I did was dead simple:
* git clone zim
* modify [the bundled markdown exporter](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/formats/markdown.py)
* run from the same directory: `./zim.py --export --format markdown --output result where/is/your/notebook`

My primary changes were related to checkboxes, list formatting, and other minor stuff to make it work better with Obsidian.

A general note: don't try to make the export do everything. Changing things (like moving all attachments into a single folder, changing links, and other improvements) is much easier to do with an external script rather than bending Zim's default exports, so I did an ok markdown export suitable for Obsidian and then changed the structure using a separate script.
